sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/RetryAnalyzer.java

```diff
@@ -22,11 +24,41 @@ public RetryAnalyzer() {
     @Override
     public boolean retryMethod(ITestResult result) {
         try {
-            TimeUnit.SECONDS.sleep(waitBetweenRetriesInSeconds);
+
+            int timeToWaitBeforeRetryInSeconds = getTimeToWaitInSeconds(result);
+            TimeUnit.SECONDS.sleep(timeToWaitBeforeRetryInSeconds);
         } catch (InterruptedException e) {
             return false;
         }
 
         return true;
     }
+
+    private int getTimeToWaitInSeconds(ITestResult result) {
+        Throwable throwable = result.getThrowable();
+        CosmosClientException cosmosClientException = extractCosmosClientExceptionIfAny(throwable);
+
+        if (cosmosClientException == null) {
+            return  waitBetweenRetriesInSeconds;
+        }
+
+        long retryAfterInMilliseconds = cosmosClientException.getRetryAfterInMilliseconds();
+        if (retryAfterInMilliseconds <= 0) {
+            return waitBetweenRetriesInSeconds;
+        }
+
+        return Math.max(Math.toIntExact(Duration.ofMillis(retryAfterInMilliseconds).getSeconds()), waitBetweenRetriesInSeconds);
+    }
+
+    private CosmosClientException extractCosmosClientExceptionIfAny(Throwable t) {
+        if (t == null) {
+            return null;
+        }
+
+        if (t instanceof CosmosClientException) {
+            return (CosmosClientException) t;
+        }
+
+        return extractCosmosClientExceptionIfAny(t.getCause());
```

Comment:
Recursion should be avoided as much as we can because of obvious reasons. 
I believe this can easily be converted to a while loop.

Reply:
why?